### PR TITLE
[bsp][stm32]Optimize ADCs and fix some bugs

### PR DESCRIPTION
**Pull request by youzuwei**
## 拉取/合并请求描述：(PR description)

[

#### 为什么提交这份PR (why to submit this PR)
一、ADC 存在一些bug
1、调用adc.c文件中函数rt_int16_t rt_adc_voltage(rt_adc_device_t dev, rt_int8_t channel)获取不到正确的电压值。分析原因如下
static rt_int16_t stm32_adc_get_vref (struct rt_adc_device *device)
{
    if(device == RT_NULL)
      return -RT_ERROR;

    rt_uint16_t vref_mv;
#ifdef __LL_ADC_CALC_VREFANALOG_VOLTAGE
    rt_err_t ret = RT_EOK;
    rt_uint32_t vref_value;

    ADC_HandleTypeDef *stm32_adc_handler = device->parent.user_data;

    ret = rt_adc_enable(device, RT_ADC_INTERN_CH_VREF);
    if (ret != RT_EOK) return (rt_int16_t)ret; // 错误的时候，这里返回ret
    vref_value = rt_adc_read(device, RT_ADC_INTERN_CH_VREF);
    ret = rt_adc_disable(device, RT_ADC_INTERN_CH_VREF); // 会停止ADC外设
    if (ret != RT_EOK) return (rt_int16_t)ret;

#ifdef SOC_SERIES_STM32U5
    vref_mv = __LL_ADC_CALC_VREFANALOG_VOLTAGE(stm32_adc_handler->Instance, vref_value, stm32_adc_handler->Init.Resolution);
#else
    vref_mv = __LL_ADC_CALC_VREFANALOG_VOLTAGE(vref_value, stm32_adc_handler->Init.Resolution);
#endif

#else
    vref_mv = 3300;
#endif /* __LL_ADC_CALC_VREFANALOG_VOLTAGE */
    return vref_mv; //正确返回
}

static rt_err_t _adc_control(rt_device_t dev, int cmd, void *args)
{
    rt_err_t result = -RT_EINVAL;
    rt_adc_device_t adc = (struct rt_adc_device *)dev;

    if (cmd == RT_ADC_CMD_ENABLE && adc->ops->enabled)
    {
        result = adc->ops->enabled(adc, (rt_uint32_t)args, RT_TRUE);
    }
    else if (cmd == RT_ADC_CMD_DISABLE && adc->ops->enabled)
    {
        result = adc->ops->enabled(adc, (rt_uint32_t)args, RT_FALSE);
    }
    else if (cmd == RT_ADC_CMD_GET_RESOLUTION && adc->ops->get_resolution && args)
    {
        rt_uint8_t resolution = adc->ops->get_resolution(adc);
        if(resolution != 0)
        {
            *((rt_uint8_t*)args) = resolution;
            LOG_D("resolution: %d bits", resolution);
            result = RT_EOK;
        }
    }
    else if (cmd == RT_ADC_CMD_GET_VREF && adc->ops->get_vref && args)
    {
        rt_int16_t value = adc->ops->get_vref(adc); // 如果函数stm32_adc_get_vref()转换错误返回是一个负值
        if(value != 0) // 负值不等于0
        {
            *((rt_int16_t *) args) = value; // 这里将是一个错误的数据
            result = RT_EOK;
        }
    }

    return result;
}

rt_int16_t rt_adc_voltage(rt_adc_device_t dev, rt_int8_t channel)
{
    rt_uint32_t value = 0;
    rt_int16_t vref = 0, voltage = 0;
    rt_uint8_t resolution = 0;

    RT_ASSERT(dev);

    /*get the resolution in bits*/
    if (_adc_control((rt_device_t) dev, RT_ADC_CMD_GET_RESOLUTION, &resolution) != RT_EOK)
    {
        goto _voltage_exit;
    }

    /*get the reference voltage*/
    // 根据上边的分析，如果上边返回一个错误的数据，这里也将接收到，并且返回也为RT_EOK
    if (_adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, &vref) != RT_EOK) // 函数后会关闭ADC外设
    // 所以运行这个函数后，不管返回值正不正确到后边都读取不到正确的通道电压
    {
        goto _voltage_exit;
    }

    /*read the value and convert to voltage*/
    dev->ops->convert(dev, channel, &value); // 不会开启ADC外设，读取不到通道数据，且没有错误返回判断
    voltage = value * vref / ((1 << resolution) - 1);

_voltage_exit:
    return voltage;
}

2、函数static rt_ssize_t _adc_read(rt_device_t dev, rt_off_t pos, void *buffer, rt_size_t size)中形参pos应该是转换通道号，再该函数内转换一次后，通道号加size(int)，这里应该是错误的
static rt_ssize_t _adc_read(rt_device_t dev, rt_off_t pos, void *buffer, rt_size_t size)
{
    rt_err_t result = RT_EOK;
    rt_size_t i;
    struct rt_adc_device *adc = (struct rt_adc_device *)dev;
    rt_uint32_t *value = (rt_uint32_t *)buffer;

    for (i = 0; i < size; i += sizeof(int)) 
    {
        result = adc->ops->convert(adc, pos + i, value);
        if (result != RT_EOK)
        {
            return 0;
        }
        value++;
    }

    return i;
}

二、ADC优化
每次调用函数stm32_adc_get_value()获取值的时候，都需要先调用函数stm32_adc_enabled()使能adc。这对于需要使能adc后需要多次读取同一通道的数据存在效率低下的问题。

#### 请提供验证的bsp和config (provide the config and bsp) 

- BSP: bsp/stm32/stm32h743-atk-apollo

- .config:

- action:

]

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [x] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [ ] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 代码是高质量的 Code in this PR is of high quality
- [x] 已经使用[formatting](https://github.com/mysterywolf/formatting) 等源码格式化工具确保格式符合[RT-Thread代码规范](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_en.md)

### CODE REVIEW

**Comment by mysterywolf** [comment]
CI报错与此PR无关 请无视

**Review by Rbb666 — APPROVED** [review]

**Comment by Rbb666** [comment]
@youzuwei 需要解决下冲突

**Comment by youzuwei** [comment]
> @youzuwei 需要解决下冲突

**Comment by Rbb666** [comment]
@youzuwei 没事了，我帮你变基好了

**Comment by Rbb666** [comment]
@youzuwei 麻烦把具体修复了什么bug在PR中描述下

**Comment by youzuwei** [comment]
以描述，不知道理解是否正确

File: components/drivers/misc/adc.c [hunk 1]
```diff
@@ -291,6 +296,6 @@ static int adc(int argc, char **argv)
     }
     return RT_EOK;
 }
-MSH_CMD_EXPORT(adc, adc [option]);
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:294]: 这里空格应该保留

File: components/drivers/misc/adc.c [hunk 2]
```diff
@@ -291,6 +296,6 @@ static int adc(int argc, char **argv)
...
+MSH_CMD_EXPORT(adc, adc[option]);
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:299]:
```suggestion
MSH_CMD_EXPORT(adc, adc [option]);
```

**Review by mysterywolf — APPROVED** [review]

**Comment by cha331** [comment]
代码格式问题很多啊，这也能merge
image

**Comment by JonasWen** [comment]
这个PR咋回事，改的太不谨慎了吧

**Pull request merged**